StyleCop.Analyzers/StyleCop.Analyzers/MaintainabilityRules/SA1402FileMayOnlyContainASingleClass.cs

```diff
@@ -60,37 +62,35 @@ private static void HandleCompilationStart(CompilationStartAnalysisContext conte
         private static void HandleSyntaxTree(SyntaxTreeAnalysisContext context)
         {
             var syntaxRoot = context.Tree.GetRoot(context.CancellationToken);
-
             var descentNodes = syntaxRoot.DescendantNodes(descendIntoChildren: node => node != null && !node.IsKind(SyntaxKind.ClassDeclaration));
+            var classNodes = from descentNode in descentNodes
+                             where descentNode.IsKind(SyntaxKind.ClassDeclaration)
+                             select descentNode as ClassDeclarationSyntax;
 
-            string foundClassName = null;
-            bool isPartialClass = false;
+            var preferredClassNode = classNodes.FirstOrDefault(n => n.Identifier.Text == Path.GetFileNameWithoutExtension(context.Tree.FilePath)) ?? classNodes.FirstOrDefault();
```

Comment:
:thought_balloon: The logic for expected file names is actually contained in SA1649. With some (hopefully minor) refactoring, we could reuse that logic here since it accounts for the preferred file naming convention.